File: python/visitor/visitor.py
import math
from parsing.AsaliLangGrammarLexer import AsaliLangGrammarLexer
from parsing.AsaliLangGrammarVisitor import AsaliLangGrammarVisitor
from parsing.AsaliLangGrammarParser import AsaliLangGrammarParser
from helpers.methods import Methods


class CustomVisitor(AsaliLangGrammarVisitor):
    def __init__(self) -> None:
        super().__init__()
        self.vars = {}
        self.methods = Methods()

    def visitParse(self, ctx):
        self.visit(ctx.block())

    def visitBlock(self, ctx):
        for stat in ctx.stat():
            self.visit(stat)

    def visitStatBlock(self, ctx):
        if ctx.stat() != None:
            return self.visit(ctx.stat())
        elif ctx.block() != None:
            return self.visit(ctx.block())
        return None

    def visitAssignment(self, ctx):
        value = self.visit(ctx.expr())
        if str(value).isnumeric():
            value = float(value)
        self.vars[ctx.ID().getText()] = value
        return super().visitAssignment(ctx)

    def visitMethodCallExpr(self, ctx):
        return self.visit(ctx.inlineMethodCall())

    def visitPowExpr(self, ctx):
        left = self.visit(ctx.expr(0))
        right = self.visit(ctx.expr(1))
        return math.pow(left, right)

    def visitUnaryMinusExpr(self, ctx):
        return -1 * self.visit(ctx.expr())

    def visitMultiplicationExpr(self, ctx):
        left = self.visit(ctx.expr(0))
        right = self.visit(ctx.expr(1))
        op = ctx.op.type
        if op == AsaliLangGrammarLexer.MULT:
            return left * right
        elif op == AsaliLangGrammarLexer.DIV:
            return left / right
        else:
            return left % right

    def visitAdditiveExpr(self, ctx):
        left = self.visit(ctx.expr(0))
        right = self.visit(ctx.expr(1))
        op = ctx.op.type
        if not isinstance(left, float) or not isinstance(right, float):
            return str(left) + str(right)
        if op == AsaliLangGrammarLexer.PLUS:
            return left + right
        else:
            return left - right

    def visitRelationalExpr(self, ctx):
        left = self.visit(ctx.expr(0))
        right = self.visit(ctx.expr(1))
        op = ctx.op.type
        if op == AsaliLangGrammarLexer.LT:
            return left < right
        elif op == AsaliLangGrammarLexer.LTEQ:
            return left <= right
        elif op == AsaliLangGrammarLexer.GT:
            return left > right
        else:
            return left >= right

    def visitEqualityExpr(self, ctx):
        left = self.visit(ctx.expr(0))
        right = self.visit(ctx.expr(1))
        op = ctx.op.type
        if op == AsaliLangGrammarLexer.EQ:
            return left == right
        elif op == AsaliLangGrammarLexer.NEQ:
            return left != right

    def visitAndExpr(self, ctx):
        left = self.visit(ctx.expr(0))
        right = self.visit(ctx.expr(1))
        return left and right

    def visitOrExpr(self, ctx):
        left = self.visit(ctx.expr(0))
        right = self.visit(ctx.expr(1))
        return left or right

    def visitNotExpr(self, ctx):
        return not self.visit(ctx.expr())

    def visitParExpr(self, ctx):
        return self.visit(ctx.expr())

    def visitAtomExpr(self, ctx):
        return self.visit(ctx.atom())

    def visitIdAtom(self, ctx):
        re = self.vars[ctx.ID().getText()]
        return re

    def visitStringAtom(self, ctx):
        value = ctx.getText()
        value: str = value[1:len(value) - 1]
        return value.replace("\"\"", "\"")

    def visitNumberAtom(self, ctx):
        value = ctx.getText()
        return float(value)

    def visitBooleanAtom(self, ctx):
        return ctx.TRUE() != None

    def visitMethodCall(self, ctx):
        return self.handleMethodCall(ctx)

    def visitInlineMethodCall(self, ctx):
        return self.handleMethodCall(ctx)

    def handleMethodCall(self, ctx):
        args = self.visit(ctx.methodCallArguments())
        methodName = ctx.ID().getText()
        if methodName == "print":
            self.methods.print(args)
        if methodName == "sin":
            return self.methods.sin(args)
        return None

    def visitMethodCallArguments(self, ctx):
        exprList = ctx.expr()
        result = []
        for expr in exprList:
            result.append(self.visit(expr))
        return result

    def visitMethodCallStat(self, ctx):
        return self.visit(ctx.methodCall())

    def visitForStat(self, ctx):
        varName = ctx.ID().getText()
        start = int(self.visit(ctx.expr(0)))
        end = int(self.visit(ctx.expr(1)))

        # Last variable value for this name
        self.vars[varName] = start

        # for i in range(start, end):
        while self.vars[varName] < end:
            self.visit(ctx.statBlock())
            self.vars[varName] = self.vars[varName] + 1

    def visitLoopStat(self, ctx):
        varName = ctx.ID().getText()
        end = int(self.visit(ctx.expr()))

        self.vars[varName] = 1
        while self.vars[varName] <= end:
            self.visit(ctx.statBlock())
            self.vars[varName] = self.vars[varName] + 1

    def visitWhileStat(self, ctx):
        result = self.visit(ctx.expr())
        while result:
            self.visit(ctx.statBlock())
            result = self.visit(ctx.expr())

    def visitIfStat(self, ctx):
        conditions = ctx.conditionBlock()
        evaluatedBlock = False
        for condition in conditions:
            evaluated = self.visit(condition.expr())
            if evaluated:
                evaluatedBlock = True
                self.visit(condition.statBlock())
                break

        if not evaluatedBlock:
            if ctx.statBlock():
                self.visit(ctx.statBlock())

File: go/project/visitor.go
package project

import (
	"fmt"
	"math"
	"strconv"
	"strings"

	parsing "github.com/abolfazlalz/AsaliLang/parsing"
	"github.com/antlr4-go/antlr/v4"
)

type Visitor struct {
	*parsing.BaseAsaliLangGrammarVisitor
	method *BuildInMethod
	vars   map[string]interface{}
}

func NewVisitor() *Visitor {
	return &Visitor{
		vars: map[string]interface{}{
			"PI": math.Pi,
		},
		method:                      NewBuildInMethod(),
		BaseAsaliLangGrammarVisitor: &parsing.BaseAsaliLangGrammarVisitor{},
	}
}

func (v *Visitor) Visit(tree antlr.ParseTree) interface{} {
	return tree.Accept(v)
}

func (v *Visitor) VisitBlock(ctx *parsing.BlockContext) interface{} {
	for _, c := range ctx.AllStat() {
		v.Visit(c)
	}
	return nil
}

func (v *Visitor) VisitStatBlock(ctx *parsing.StatBlockContext) interface{} {
	if ctx.Stat() != nil {
		v.Visit(ctx.Stat())
		return nil
	}
	for _, c := range ctx.Block().AllStat() {
		v.Visit(c)
	}
	return nil
}

func (v *Visitor) VisitParse(ctx *parsing.ParseContext) interface{} {
	v.Visit(ctx.Block())
	return nil
}
func (v *Visitor) VisitStat(ctx *parsing.StatContext) interface{} {
	return v.Visit(ctx.GetChild(0).(antlr.ParseTree))
}

func (v *Visitor) VisitAssignment(ctx *parsing.AssignmentContext) interface{} {
	v.vars[ctx.ID().GetText()] = v.Visit(ctx.Expr())
	return nil
}

func (v *Visitor) VisitMethodCallExpr(ctx *parsing.MethodCallExprContext) interface{} {
	return v.Visit(ctx.InlineMethodCall())
}

func (v *Visitor) VisitPowExpr(ctx *parsing.PowExprContext) interface{} {
	left := v.Visit(ctx.Expr(0))
	right := v.Visit(ctx.Expr(1))
	return math.Pow(left.(float64), right.(float64))
}

func (v *Visitor) VisitUnaryMinusExpr(ctx *parsing.UnaryMinusExprContext) interface{} {
	result := v.Visit(ctx.Expr())
	if value, ok := result.(float64); ok {
		return value * -1
	} else if value, ok := result.(int); ok {
		return value * -1
	} else if value, ok := result.(int64); ok {
		return value * -1
	}
	panic(fmt.Sprintf("Value to minus must be integer or float"))
}

func (v *Visitor) VisitMultiplicationExpr(ctx *parsing.MultiplicationExprContext) interface{} {
	left := v.Visit(ctx.Expr(0)).(float64)
	right := v.Visit(ctx.Expr(1)).(float64)

	switch ctx.GetOp().GetTokenType() {
	case parsing.AsaliLangGrammarLexerMULT:
		return left * right
	case parsing.AsaliLangGrammarLexerDIV:
		return left / right
	case parsing.AsaliLangGrammarLexerMOD:
		return math.Mod(left, right)
	}

	panic("Undefined operator")
}

func (v *Visitor) VisitAdditiveExpr(ctx *parsing.AdditiveExprContext) interface{} {
	left := v.Visit(ctx.Expr(0))
	right := v.Visit(ctx.Expr(1))

	switch ctx.GetOp().GetTokenType() {
	case parsing.AsaliLangGrammarLexerPLUS:
		valLeft, okLeft := left.(float64)
		valRight, okRight := right.(float64)
		if okRight && okLeft {
			return valRight + valLeft
		}
		return fmt.Sprintf("%v", left) + fmt.Sprintf("%v", right)
	case parsing.AsaliLangGrammarLexerMINUS:
		return left.(float64) - right.(float64)
	}
	panic("Undefined operator")
}

func (v *Visitor) VisitRelationalExpr(ctx *parsing.RelationalExprContext) interface{} {
	left := v.Visit(ctx.Expr(0)).(float64)
	right := v.Visit(ctx.Expr(1)).(float64)

	switch ctx.GetOp().GetTokenType() {
	case parsing.AsaliLangGrammarLexerLT:
		return left < right
	case parsing.AsaliLangGrammarLexerLTEQ:
		return left <= right
	case parsing.AsaliLangGrammarLexerGT:
		return left > right
	case parsing.AsaliLangGrammarLexerGTEQ:
		return left >= right
	}
	panic("Undefined operator")
}

func (v *Visitor) VisitEqualityExpr(ctx *parsing.EqualityExprContext) interface{} {
	left := v.Visit(ctx.Expr(0))
	right := v.Visit(ctx.Expr(1))

	switch ctx.GetOp().GetTokenType() {
	case parsing.AsaliLangGrammarLexerEQ:
		return left == right
	case parsing.AsaliLangGrammarLexerNEQ:
		return left != right
	}
	panic("Undefined operator")
}

func (v *Visitor) VisitAndExpr(ctx *parsing.AndExprContext) interface{} {
	left := newValue(v.Visit(ctx.Expr(0))).asBoolean()
	right := newValue(v.Visit(ctx.Expr(1))).asBoolean()
	return right && left
}

func (v *Visitor) VisitOrExpr(ctx *parsing.OrExprContext) interface{} {
	left := newValue(v.Visit(ctx.Expr(0))).asBoolean()
	right := newValue(v.Visit(ctx.Expr(1))).asBoolean()
	return right || left
}

func (v *Visitor) VisitAtomExpr(ctx *parsing.AtomExprContext) interface{} {
	return v.Visit(ctx.Atom())
}

func (v *Visitor) VisitIdAtom(ctx *parsing.IdAtomContext) interface{} {
	return v.vars[ctx.ID().GetText()]
}

func (v *Visitor) VisitStringAtom(ctx *parsing.StringAtomContext) interface{} {
	str := ctx.GetText()
	str = str[1 : len(str)-1]
	str = strings.ReplaceAll(str, "\"\"", "\"")
	return str
}

func (v *Visitor) VisitNotExpr(ctx *parsing.NotExprContext) interface{} {
	value := v.Visit(ctx.Expr())
	return !newValue(value).asBoolean()
}

func (v *Visitor) VisitParExpr(ctx *parsing.ParExprContext) interface{} {
	return v.Visit(ctx.Expr())
}

func (v *Visitor) VisitNumberAtom(ctx *parsing.NumberAtomContext) interface{} {
	if ctx.INT() != nil {
		val, _ := strconv.ParseFloat(ctx.INT().GetText(), 64)
		return val
	} else if ctx.FLOAT() != nil {
		val, _ := strconv.ParseFloat(ctx.FLOAT().GetText(), 64)
		return val
	}
	return nil
}

func (v *Visitor) VisitBooleanAtom(ctx *parsing.BooleanAtomContext) interface{} {
	return ctx.TRUE() != nil
}

func (v *Visitor) VisitMethodCall(ctx *parsing.MethodCallContext) interface{} {
	methodName := ctx.ID().GetText()
	return v.handleMethodCall(methodName, ctx.MethodCallArguments())
}

func (v *Visitor) VisitInlineMethodCall(ctx *parsing.InlineMethodCallContext) interface{} {
	id := ctx.ID().GetText()
	return v.handleMethodCall(id, ctx.MethodCallArguments())
}

func (v *Visitor) handleMethodCall(methodName string, argsContext parsing.IMethodCallArgumentsContext) interface{} {
	args := v.Visit(argsContext).([]interface{})
	if methodName == "print" {
		v.method.print(args...)
	} else if methodName == "sin" {
		return v.method.sin(args...)
	}
	return nil
}

func (v *Visitor) VisitMethodCallArguments(ctx *parsing.MethodCallArgumentsContext) interface{} {
	exprList := ctx.AllExpr()
	result := make([]interface{}, len(exprList))
	for i, exprList := range exprList {
		result[i] = v.Visit(exprList)
	}
	return result
}

func (v *Visitor) VisitMethodCallStat(ctx *parsing.MethodCallStatContext) interface{} {
	return v.Visit(ctx.MethodCall())
}

func (v *Visitor) VisitForStat(ctx *parsing.ForStatContext) interface{} {
	varName := ctx.ID().GetText()
	start := newValue(v.Visit(ctx.Expr(0))).asFloat()
	end := newValue(v.Visit(ctx.Expr(1))).asFloat()
	lastVariable := v.vars[varName]
	for v.vars[varName] = start; toFloat(v.vars[varName]) < end; {
		v.Visit(ctx.StatBlock())
		v.vars[varName] = toFloat(v.vars[varName]) + 1
	}
	v.vars[varName] = lastVariable
	return nil
}

func (v *Visitor) VisitLoopStat(ctx *parsing.LoopStatContext) interface{} {
	varName := ctx.ID().GetText()
	end := newValue(v.Visit(ctx.Expr())).asFloat()

	lastVariable := v.vars[varName]

	for v.vars[varName] = 0; toFloat(v.vars[varName]) < end; {
		v.vars[varName] = toFloat(v.vars[varName]) + 1
		v.Visit(ctx.StatBlock())
	}

	v.vars[varName] = lastVariable

	return nil
}

func (v *Visitor) VisitWhileStat(ctx *parsing.WhileStatContext) interface{} {
	result := toBoolean(v.Visit(ctx.Expr()))
	for result {
		result = toBoolean(v.Visit(ctx.Expr()))
		v.Visit(ctx.StatBlock())
	}
	return nil
}

func (v *Visitor) VisitIfStat(ctx *parsing.IfStatContext) interface{} {
	conditions := ctx.AllConditionBlock()
	evaluatedBlock := false
	for _, condition := range conditions {
		evaluated := v.Visit(condition.Expr())
		if toBoolean(evaluated) {
			evaluatedBlock = true
			v.Visit(condition.StatBlock())
			break
		}
	}

	if !evaluatedBlock && ctx.StatBlock() != nil {
		v.Visit(ctx.StatBlock())
	}

	return nil
}
